quicktill/pdrivers.py

```diff
@@ -701,6 +716,38 @@ def process_canvas(self, canvas, f):
                 f.write(escpos.ep_fullcut)
             f.flush()
 
+    def _image(self, data, width, height, f):
+        # Print a PBM bit-image
+        if width > self.dpl:
+            # Image too wide for paper
+            return
+
+        # Partition the bitmap into rows
+        rows = []
+        for line in range(height):
+            start = line * (width // 8)
+            end = start + (width // 8)
+            rows.append(data[start:end])
+
+        # Calculate padding required to center the image
+        padding = (self.dpl - width) // 2
+        padchars = [False] * padding
+
+        # Write the commands to render the padded image
+        f.write(escpos.ep_unidirectional_on)
+        for row in rows:
+            line = padchars.copy()
```

Comment:
I think you need to be writing 8 rows at a time to the printer, since you're using the ESC/POS command that sends 8 "pins" worth of data at once.

So grab 8 rows of raw image data, rotate it so that you are sending columns of 8 pixels, and repeat until you run out of rows.

Classic AoC Python hint: if you have `[("a", "b", "c"), ("d", "e", "f")]` and you want `[("d", "a"), ("e", "b"), ("f", "c")]` (i.e. clockwise rotation) you should be using `list(zip(*reversed(the_thing)))`

Reply:
Brilliant, thank you - yep, that sounds like exactly what's needed.  Better results soon, hopefully :)